File: model/vote.rb
require_relative 'application_model'
require_relative 'vote_result'

class Vote < ApplicationModel
  attr_reader :id, :bill_id

  def self.filename
    'votes'
  end

  def self.all
    read.map { |row| new(row[:id], row[:bill_id]) }
  end

  def initialize(id, bill_id)
    super()
    @id = id
    @bill_id = bill_id
  end

  def results
    @results = VoteResult.filter_by(column: :vote_id, value: id)
  end

  def supports
    results.select(&:support?)
  end

  def oppositions
    results.select(&:opposed?)
  end
end

File: model/vote.js
import ApplicationModel from "./application_model.js";
import VoteResult from "./vote_result.js";

export default class Vote extends ApplicationModel {
  static filename = "votes";

  static async all() {
    return this.read().then((array) =>
      array.map((row) => new this(row.id, row.bill_id))
    );
  }

  constructor(id, bill_id) {
    super();
    this.id = id;
    this.bill_id = bill_id;
  }

  async results() {
    return await VoteResult.filter_by('vote_id', this.id);
  }

  async supports() {
    const results = await this.results();
    return results.filter((result) => result.support());
  }

  async oppositions() {
    const results = await this.results();
    return results.filter((result) => result.opposed());
  }
}
